linkerd/proxy/discover/src/from_resolve.rs

```diff
@@ -163,50 +185,112 @@ where
 #[cfg(test)]
 mod tests {
     use super::Discover;
-    use async_stream::stream;
     use futures::prelude::*;
     use linkerd_error::Infallible;
     use linkerd_proxy_core::resolve::Update;
     use std::net::SocketAddr;
+    use tokio_stream::wrappers::ReceiverStream;
     use tower::discover::Change;
 
     const PORT: u16 = 8080;
     fn addr(n: u8) -> SocketAddr {
         SocketAddr::from(([10, 1, 1, n], PORT))
     }
 
-    #[tokio::test]
+    #[tokio::test(flavor = "current_thread")]
     async fn reset() {
-        tokio::pin! {
-            let stream = stream! {
-                yield Ok::<_, Infallible>(Update::Add((1..=2).map(|n| (addr(n), n)).collect()));
-                yield Ok(Update::Reset((2..=4).map(|n| (addr(n), n)).collect()));
-            };
-        }
-        let mut disco = Discover::new(stream);
-
-        for i in 1..=2 {
-            match disco.next().await.unwrap().unwrap() {
-                Change::Remove(_) => panic!("Unexpectd Remove"),
-                Change::Insert(a, n) => {
-                    assert_eq!(n, i);
-                    assert_eq!(addr(i), a);
-                }
-            }
-        }
-        match disco.next().await.unwrap().unwrap() {
-            Change::Remove(a) => assert_eq!(a, addr(1)),
-            change => panic!("Unexpected change: {:?}", change),
+        let (tx, rx) = tokio::sync::mpsc::channel(1);
+        let mut disco = Discover::new(ReceiverStream::new(rx));
+
+        // Use reset to set a new state with 3 addresses.
+        tx.try_send(Ok::<_, Infallible>(Update::Reset(
+            (1..=3).map(|n| (addr(n), n)).collect(),
+        )))
+        .expect("must send");
+        for i in 1..=3 {
+            assert!(matches!(
+                disco.try_next().await,
+                Ok(Some(Change::Insert(sa, n))) if sa == addr(i) && n == i
+            ));
         }
+
+        // Reset to a new state with 3 addresses, one of which is unchanged, one of which is
+        // changed, and one of which is added.
+        tx.try_send(Ok(Update::Reset(
+            // Restore the original `2` value. We shouldn't see an update for it.
+            Some((addr(2), 2))
+                .into_iter()
+                // Set a new value for `3`. and add a 4th as well.
+                .chain((3..=4).map(|n| (addr(n), n + 1)))
+                .collect(),
```

Comment:
tbh, i find this a little weird, it might be clearer to just write
```suggestion
            vec![
                // Restore the original `2` value. We shouldn't see an update for it.
                (addr(2), 2),
                // Set a new value for `3`.
                (addr(3), 4),
                // Add a fourth endpoint as well.
                (addr(4), 5),
            ]
```